### PR TITLE
Add `Expr` support to `ControlFlowOp` representation

### LINKED ISSUE #10224: Add support for classical expressions to `SwitchCaseOp`, `WhileLoopOp` and `IfElseOp`
### What should we add?

This tracks the work labelled "PR 2" in [the initial classical-expression design document](https://github.com/Qiskit/RFCs/blob/master/0010-simple-classical-representations.md).

Depends on #10223.

Brief overview:
- relax the type checking to permit `WhileLoopOp.condition` and `IfElseOp.condition` to be `Expr` values with type `Bool`, and `SwitchCaseOp.target` to be an `Expr` value with type `Bool` or `Uint[n]`.

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This commit allows `IfElseOp` and `WhileLoopOp` to have a `condition` that is an `Expr` typed `Bool()`, and `SwitchCaseOp` to have a `target` that is an `Expr`.  It does not fully add support to resolving these new allowed values into the control-flow builder interface (that will come later).

### Details and comments

Close #10224.

Depends on #10332.  Same consideration about release note.

Originally I was doing the control-flow builder interface as part of this commit as well since there's very little additional work needed to make that work.  In practice, though, being able to _test_ the builder interface requires a bunch of the `DAGCircuit`-related bits to be working as well, so I've split it off.

### CODE REVIEW

**Review by kevinhartman — DISMISSED** [review]
Very straightforward. These changes include #10332, so once that merges, this one is good to go too (after a rebase/merge).

I like how you were able to reuse a lot of the existing testing, and the integration points to enable expressions in the conditions and switch target are a lot more minimal than I was expecting!

**Comment by jakelishman** [comment]
Now rebased on top of `main`.

**Review by kevinhartman — APPROVED** [review]

**Pull request merged**
